File: sdks/go/raisin/host.go
// SPDX-License-Identifier: BSL-1.1
//
// RaisinDB - Git-like hierarchical multi model database
// Copyright (C) 2019-2025 SOLUTAS GmbH, Switzerland

package raisin

import "sync"

// LogLevel mirrors the WIT `log-level` enum.
type LogLevel uint8

// The log levels understood by the host.
const (
	LevelDebug LogLevel = iota
	LevelInfo
	LevelWarn
	LevelError
)

// String returns the lowercase level name used by the host log surface.
func (l LogLevel) String() string {
	switch l {
	case LevelDebug:
		return "debug"
	case LevelInfo:
		return "info"
	case LevelWarn:
		return "warn"
	case LevelError:
		return "error"
	default:
		return "info"
	}
}

// Host is the RaisinDB host surface, one method per WIT import.
//
// There is exactly one implementation per build target: the component-model
// bindings under `wasip2`/`tinygo`, and whatever a test installs natively.
// Guest code never talks to a Host directly — it calls the generated
// namespaces, which funnel through the call helpers in call.go.
type Host interface {
	// Call invokes a RaisinDB API method by registry internal name with a
	// JSON array of positional arguments, returning the raw JSON payload.
	Call(method string, args string) (string, error)
	// Log emits a structured log line into the execution result.
	Log(level LogLevel, message string)
	// Context returns the execution-context JSON, byte-identical to
	// `raisin.context.get()` in JavaScript and Starlark.
	Context() string
	// ABIVersion returns the host ABI semver, e.g. "0.1.0".
	ABIVersion() string
}

var (
	hostMu      sync.RWMutex
	installed   Host
	defaultHost Host
)

// SetHost installs a Host, returning a function that restores the previous
// one. Tests use it through raisintest; guest code never calls it.
func SetHost(h Host) (restore func()) {
	hostMu.Lock()
	prev := installed
	installed = h
	hostMu.Unlock()
	return func() {
		hostMu.Lock()
		installed = prev
		hostMu.Unlock()
	}
}

// currentHost returns the installed host, falling back to the build target's
// default (the component bindings on wasm, a failing stub natively).
func currentHost() Host {
	hostMu.RLock()
	h := installed
	hostMu.RUnlock()
	if h != nil {
		return h
	}
	return defaultHost
}

// ABIVersion reports the host ABI version this component is running against.
func ABIVersion() string {
	return currentHost().ABIVersion()
}

File: sdks/ts/function-wasm/src/host.js
// SPDX-License-Identifier: BSL-1.1
//
// RaisinDB - Git-like hierarchical multi model database
// Copyright (C) 2019-2025 SOLUTAS GmbH, Switzerland

/**
 * The WIT host, held in one place.
 *
 * The real host arrives as a Component Model import (`raisin:function/host`),
 * which only resolves inside `jco componentize`. Node — and therefore vitest —
 * cannot resolve that specifier at all, so the import lives in its own module
 * (`host-wit.js`, pulled in only by the generated component entry) and
 * everything else in this SDK goes through the holder below. That is what lets
 * the whole surface be tested natively against a mock host.
 *
 * @typedef {object} RaisinHost
 * @property {(method: string, args: string) => string} call
 *   Registry gateway. `args` is a JSON array of positional arguments; the
 *   return is `InvokeResult::to_json_string()`. A host-side failure THROWS
 *   (the Component Model `result<string, string>` err arm).
 * @property {(level: 'debug'|'info'|'warn'|'error', message: string) => void} log
 * @property {() => string} context   Execution context, as JSON.
 * @property {() => string} abiVersion Host ABI semver, e.g. "0.1.0".
 */

/** The ABI version this SDK was generated against. */
export const SDK_ABI_VERSION = '0.1.0';

/** @type {RaisinHost | null} */
let current = null;

/**
 * Install the host implementation. Called once by `host-wit.js` in a real
 * component, and by tests with a mock.
 *
 * @param {RaisinHost} host
 */
export function setHost(host) {
  if (!host || typeof host.call !== 'function') {
    throw new TypeError('setHost() needs a host with a call(method, args) function');
  }
  current = host;
}

/** Forget the installed host. Test-only; a component never calls it. */
export function clearHost() {
  current = null;
}

/**
 * The installed host.
 *
 * @returns {RaisinHost}
 */
export function host() {
  if (!current) {
    throw new Error(
      'no RaisinDB host is installed: a component entry must import ' +
        "'@raisindb/function-wasm/host-wit', a test must call setHost(mockHost)",
    );
  }
  return current;
}

/** True when a host is installed. */
export function hasHost() {
  return current !== null;
}
